### PR TITLE
Add thunderbird policies

### PR DESCRIPTION
**Pull request by nydragon**
## Description of changes

Addition of the thunderbird policies and about:config options to reflect firefox's capabilities and to allow system wide locking of options.

This is in large parts inspired from the firefox module.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/thunderbird.nix
Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:10]: The issue is that the documentation for the policies do not match the Thunderbird version 1:1. [Here](https://github.com/thunderbird/policy-templates/tree/master/templates/) they have a folder for esr102 esr115 est68 esr78 esr91 and central (the most up to date one).

I could just set the link to point to the most [recent documentation](https://github.com/thunderbird/policy-templates/tree/master/templates/central) at a higher risk that policies are misdocumented

Comment (Kreyren) [reply nixos/modules/programs/thunderbird.nix:10]: Could we ask upstream to make a better QA for this so that the downstream doesn't have to do management like this?

Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:10]: They do not have issues activated in their repository and I doubt they would make changes to accommodate us.

The Firefox package links to the most up to date version of the documentation regardless of Firefox version, so I suppose we could just link to the [github repository root](https://github.com/thunderbird/policy-templates) where a user can select their version or to [enterprise.thunderbird.net/manage-updates-policies-and-customization](https://enterprise.thunderbird.net/manage-updates-policies-and-customization/managing-thunderbird-policies/customizing-thunderbird-using-policies.json), which in turn links to GitHub.

I'll write them a support ticket making this request, but this will take some time to happen, if at all.

Comment (Kreyren) [reply nixos/modules/programs/thunderbird.nix:10]: > I doubt they would make changes to accommodate us. -- @Nydragon(https://github.com/NixOS/nixpkgs/pull/311669#discussion_r1601695988)

This rather seems as release management issue rather than NixOS-specific thing to me, thus why i think it might be a good idea to point that out to them.

> The Firefox package links to the most up to date version of the documentation regardless of Firefox version, so I suppose we could just link to the [github repository root](https://github.com/thunderbird/policy-templates) where a user can select their version or to [enterprise.thunderbird.net/manage-updates-policies-and-customization](https://enterprise.thunderbird.net/manage-updates-policies-and-customization/managing-thunderbird-policies/customizing-thunderbird-using-policies.json), which in turn links to GitHub. -- @Nydragon (https://github.com/NixOS/nixpkgs/pull/311669#discussion_r1601695988)

My point is that the less things we need to constantly change on the packages the better as then the development time gets progressively better to focus on other issues and without introducing confusion to the end-user who might experience a situation where they get a wrong version of the documentation.

> I'll write them a support ticket making this request, but this will take some time to happen, if at all. -- @Nydragon (https://github.com/NixOS/nixpkgs/pull/311669#discussion_r1601695988)

Reference the reply if you get any plz might be also a good idea to reference this in the code itself

Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:10]: I reached out on [Matrix](https://matrix.to/#/!fubIsJzeAcCcjYTQvm:mozilla.org/$4byxRrFJmBIJRIm9azPT-jVF6HmSjBCNkaetM5Qip2E?via=kewis.ch&via=mozilla.org&via=matrix.org) and their [support site](https://support.mozilla.org/en-US/questions/1447457), i'll update here once I get a reply.

> the less things we need to constantly change on the packages the better

I fully agree, but wouldn't it make sense to change the documentation URL in a subsequent PR instead of blocking this one?

Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:10]: Didn't really amount to much, you can look at the Matrix conversation for details.

So we could either point to the repo root, which links to the different ESR versions of the doc or write a small script that selects the previous ESR version based on the packaged Thunderbird version and builds the link.

I think linking to the repo root might be more reliable

File: nixos/modules/programs/thunderbird.nix
Comment (nevivurn) [inline nixos/modules/programs/thunderbird.nix:83]: Nit, we can set `source = policiesJSON`, no need to interpolate to string here.

File: nixos/modules/programs/thunderbird.nix
Comment (nevivurn) [inline nixos/modules/programs/thunderbird.nix:86]: I think this should be shown in the options documentation, we can let `programs.thunderbird.policies` be a freeform submodule and explicitly add `DisableAppUpdate` with `default = false;`.

File: nixos/modules/programs/thunderbird.nix
Comment (nevivurn) [inline nixos/modules/programs/thunderbird.nix:87]: Nit, unnecessary parens here.

**Review by nevivurn — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/thunderbird.nix
Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:86]: Couldn't this cause issues if we let Thunderbird update itself?

Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:86]: But wouldn't allowing Thunderbird to update itself cause problems?

Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:80]: https://github.com/NixOS/nixpkgs/blob/cab94ab46e676b5c6aaf98ad602d9b8ed14191dd/nixos/modules/programs/firefox.nix#L284C1-L287C7

Firefox has the same approach

Comment (nevivurn) [reply nixos/modules/programs/thunderbird.nix:80]: Ah sorry, I meant `true`. My bad.

**Review by nevivurn — APPROVED** [review]

File: nixos/modules/programs/thunderbird.nix
Comment (Nydragon) [reply nixos/modules/programs/thunderbird.nix:80]: Yeah, it's only that or `Status` with is overridable with `preferencesStatus` or manually. Thanks for the approval

**Comment by Nydragon** [comment]
@nevivurn had to resolve a merge conflict so I'd need a re-approval

**Review by nevivurn — APPROVED** [review]
No changes since my previous review

**Comment by Nydragon** [comment]
What would the next step be to get this merged?

**Comment by Nydragon** [comment]
Thank you!

**Pull request merged**
